# **Welcome to SQANTI: Structural and Quality Annotation of Novel Transcript Isoforms** #

 
###**Summary**###


SQANTI is a pipeline for the in-depth characterization of isoforms obtained by full-length transcript sequencing, which are commonly returned in a fasta file format without any extra information about gene/transcript annotation or attribute description. 

Although thought to be used for characterization of isoforms generated by long-read sequencing, SQANTI can be used for any set of transcript isoforms in fasta file format. Besides it can be applied to any organism.

SQANTI pipeline steps:

1. First, as long-read sequencing usually has a high rate of errors along sequences, SQANTI performs a **reference-based correction of sequences**.

2. Secondly, **generates genes models** and **computes a ORF prediction** for each isoform, obtaining information about the coding potential of each defined isoform. 

3. Finally it carries out a **deep characterization of isoforms at both transcript and junction level** and **generates a report** with several plots describing in detail the mayor attributes that catalog your set of sequenced-isoforms.

![Diapositiva1.png](https://bitbucket.org/repo/kpnA5g/images/69442998-Diapositiva1.png)


### **Require software** ###

* Python
* Perl
* Gmap aligner (version....)
* R environment


### **Running** ###

SQANTI is a program written in python. To run it use *sqanti.py*. Below you can see the help page of SQANTI where its mandatory and optional arguments are explained:

```
#!bash
usage: sqanti.py [-h] [-x GMAP_INDEX] [-o OUTPUT] [-d DIR] [-c COVERAGE]
                 [-s SITES] [-n] [-fl FL_COUNT] [-m] [-v]
                 isoforms ref expression genome

Structural and Quality Annotation of Novel Transcript Isoforms

positional arguments:
  isoforms              Isoforms (Fasta format)
  ref                   Reference annotation file (GTF format)
  expression            Expression matrix
  genome                Reference genome (Fasta format)

optional arguments:
  -h, --help            show this help message and exit
  -x GMAP_INDEX, --gmap_index GMAP_INDEX
                        Path and prefix of the reference index created by
                        gmap_build. Mandatory unless -m option is specified.
  -o OUTPUT, --output OUTPUT
                        Prefix for output files.
  -d DIR, --dir DIR     Directory for output files. Default: Directory where
                        the script was run.
  -c COVERAGE, --coverage COVERAGE
                        Junction coverage files (comma-separated list of
                        SJ.out.tab files generated by STAR or directory where
                        they are in).
  -s SITES, --sites SITES
                        Set of splice sites to be considered as canonical
                        (comma-separated list of splice sites). Default:
                        GTAG,GCAG,ATAC.
  -n, --name            Use gene_name tag from GTF to define genes. Default:
                        gene_id used to define genes
  -fl FL_COUNT, --fl_count FL_COUNT
                        Full-length PacBio abundance files (comma-separated
                        list of PacBio abundance files generated by PacBio or
                        directory where there are in).
  -m, --mode            Use to run Sqanti when gtf and faa files have been
                        already created in previous runs.
  -v, --version         Display program version number.

```

###**Input files**###

###**Output files**###

###**Version**###

* SQANTI 0.1


###**Contact Information**###

Lorena de la Fuente 
lfuente@cipf.es

Genomics of Gene Expression Lab
Centro de Investigación Príncipe Felipe
Eduardo Primo Yúfera, 3 46012 Valencia (Spain)

